ignition-server/controllers/tokensecret_controller.go

```diff
@@ -0,0 +1,156 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+const (
+	TokenSecretReleaseKey = "release"
+	TokenSecretTokenKey   = "token"
+	TokenSecretAnnotation = "hypershift.openshift.io/ignition-config"
+)
+
+var PayloadStore = NewPayloadStore()
+
+// PayloadsStore contains an updated map of token: payload.
+type Cache struct {
+	Map map[string][]byte
+	sync.RWMutex
+}
+
+func NewPayloadStore() *Cache {
+	return &Cache{
+		Map: make(map[string][]byte),
+	}
+}
+
+func (ps *Cache) Get(key string) (value []byte, ok bool) {
+	ps.RLock()
+	result, ok := ps.Map[key]
+	ps.RUnlock()
+	return result, ok
+}
+
+func (ps *Cache) Delete(key string) {
+	ps.Lock()
+	delete(ps.Map, key)
+	ps.Unlock()
+}
+
+func (ps *Cache) Store(key string, value []byte) {
+	ps.Lock()
+	ps.Map[key] = value
+	ps.Unlock()
+}
+
+// IgnitionProvider can build ignition payload contents
+// for a given release image.
+type IgnitionProvider interface {
+	// GetPayload returns the ignition payload contents for
+	// the provided release. In the future the list of
+	// inputs may expand to incorporate user-provided ignition
+	// overlay contents.
+	GetPayload(ctx context.Context, token *corev1.Secret) ([]byte, error)
```

Comment:
The token Secret seemed convenient since it's the contract for the controller though I see where you are coming from I updated it to be `GetPayload(ctx context.Context, releaseImage string) (payload []byte, err error) `